lib/ts/recipe/multifactorauth/recipe.ts

```diff
@@ -0,0 +1,508 @@
+                    break;
+                }
+            }
+            if (!foundVerifiedEmail) {
+                return {
+                    status: "VALIDATION_ERROR",
+                    reason: "Cannot complete MFA because of security reasons. Please contact support. (ERR_CODE_010)",
+                };
+            }
+        }
+
+        if (accountInfo.phoneNumber !== undefined) {
+            let foundVerifiedPhoneNumber = false;
+            for (const lM of sessionUser?.loginMethods) {
+                if (lM.phoneNumber === accountInfo.phoneNumber) {
+                    foundVerifiedPhoneNumber = true;
+                    break;
+                }
+            }
+            if (!foundVerifiedPhoneNumber) {
+                throw new Error("should never happen"); // phone number only comes from passwordless and is a verified factor always
+            }
+        }
+
+        return {
+            status: "OK",
+        };
+    };
+
+    checkIfFactorUserCanBeLinkedWithSessionUser = async (
+        tenantId: string,
+        sessionUser: User,
+        accountInfo: { email?: string; phoneNumber?: string },
+        userContext: UserContext
+    ): Promise<{ status: "OK" | "RECURSE_FOR_RACE" } | { status: "VALIDATION_ERROR"; reason: string }> => {
+        if (!sessionUser.isPrimaryUser) {
+            const canCreatePrimary = await AccountLinkingRecipe.getInstance().recipeInterfaceImpl.canCreatePrimaryUser({
+                recipeUserId: sessionUser.loginMethods[0].recipeUserId,
+                userContext,
+            });
+
+            if (canCreatePrimary.status === "RECIPE_USER_ID_ALREADY_LINKED_WITH_PRIMARY_USER_ID_ERROR") {
+                // Race condition since we just checked that it was not a primary user
+                this.querier.invalidateCoreCallCache(userContext);
+                return {
+                    status: "RECURSE_FOR_RACE",
+                };
+            }
+
+            if (canCreatePrimary.status === "ACCOUNT_INFO_ALREADY_ASSOCIATED_WITH_ANOTHER_PRIMARY_USER_ID_ERROR") {
+                return {
+                    status: "VALIDATION_ERROR",
+                    reason: "Cannot complete MFA because of security reasons. Please contact support. (ERR_CODE_011)",
+                };
+            }
+        }
+
+        // Check if there if the linking with session user going to fail and avoid user creation here
+        const users = await listUsersByAccountInfo(
+            tenantId,
+            { email: accountInfo.email, phoneNumber: accountInfo.phoneNumber },
+            true,
+            userContext
+        );
+        for (const user of users) {
+            if (user.isPrimaryUser && user.id !== sessionUser.id) {
+                return {
+                    status: "VALIDATION_ERROR",
+                    reason: "Cannot complete MFA because of security reasons. Please contact support. (ERR_CODE_012)",
+                };
+            }
+        }
```

Comment:
else check by calling canLinkAccount? Instead of repeating the logic here..

Reply:
can't use canLinkAccount because we don't haven't created the user yet